extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -258,6 +413,115 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
+  }
+
+  /**
+   * Executes SRA request via the following steps:
+   * - create only one executor per page
+   * - get all doubleclick amp-ad instances on the page
+   * - group by networkID allowing for separate SRA requests
+   * - for each grouping, construct SRA request
+   * - handle chunks for streaming response for each block
+   * @private
+   */
+  initiateSraRequests_() {
+    if (SraRequests) {
+      return;
+    }
+    // Use cancellation of the first slot's promiseId as indication of
+    // unlayoutCallback execution.  Assume that if called for one slot, it will
+    // be called for all and we should cancel SRA execution.
+    const checkStillCurrent = this.verifyStillCurrent();
+    SraRequests = groupAmpAdsByType(
+        this.win, this.element.getAttribute('type'), getNetworkId)
+      .then(groupIdToBlocksAry => {
+        checkStillCurrent();
+        Object.keys(groupIdToBlocksAry).forEach(networkId => {
+          const blocks = dev().assert(groupIdToBlocksAry[networkId]);
+          // TODO: filter blocks with SRA disabled?
+          return Promise.all(blocks).then(instances => {
+            dev().assert(instances.length);
+            const typeInstances =
+              /** @type {!Array<!AmpAdNetworkDoubleclickImpl>}*/(instances);
+            // Determine if more than one block for this element, if not do not
+            // set sra request promise which results in sending as
+            // non-SRA request (benefit is it allows direct cache method).
+            if (typeInstances.length == 1) {
+              dev().info(TAG, `single block in network ${networkId}`);
+              typeInstances[0].sraResponseResolver_(null);
+              return;
+            }
+            // Construct and send SRA request.
+            // Chunk hanlder called with metadata and creative for each slot
+            // in order of URLs given.  Construct promise for each slot
+            // such that its resolver will be called.
+            const sraRequestAdUrlResolvers =
+              typeInstances.map(instance => instance.sraResponseResolver_);
+            // TODO(keithwrightbos) - how do we handle per slot 204 response?
+            return constructSRARequest_(
+                this.win, this.getAmpDoc(), typeInstances)
+              .then(sraUrl => {
+                checkStillCurrent();
+                return xhrFor(this.win).fetch(sraUrl, {
+                  mode: 'cors',
+                  method: 'GET',
+                  credentials: 'include',
+                });
+              })
+              .then(response => lineDelimitedStreamer(this.win, response,
+                metaJsonCreativeGrouper((creative, headers) => {
+                  checkStillCurrent();
+                  // Force safeframe rendering method.
+                  headers[RENDERING_TYPE_HEADER] = XORIGIN_MODE.SAFEFRAME;
+                  // Need to convert to FetchResponse object?
+                  const fetchResponseHeaders =
```

Comment:
Problem is that the constructor expects !XMLHttpRequest|!XDomainRequest.  I could modify the class to take object as an optional object or can I create a mock XMLHttpRequest... thoughts?